lib_test/protocol_test.ml

```diff
@@ -12,132 +12,322 @@
+      buffers
+    |> ignore; Lwt.return_ok ()
+  let write contents sector_start buffers =
+    let sector_start = Int64.to_int sector_start in
+    Cstruct.fillv buffers (Cstruct.shift contents sector_start)
+    |> ignore; Lwt.return_ok ()
+end
+
+module V2_read_only_test = struct
+
+  let test_block = (Cstruct.of_string "asdf")
+
+  let sequence = [
+    `Server, "NBDMAGIC";
+    `Server, "IHAVEOPT";
+    `Server, "\000\001"; (* handshake flags: NBD_FLAG_FIXED_NEWSTYLE *)
+    `Client, "\000\000\000\001"; (* client flags: NBD_FLAG_C_FIXED_NEWSTYLE *)
+
+    `Client, "IHAVEOPT";
+    `Client, "\000\000\000\001"; (* NBD_OPT_EXPORT_NAME *)
+    `Client, "\000\000\000\007"; (* length of export name *)
+    `Client, "export1";
+
+    `Server, "\000\000\000\000\000\000\000\004"; (* size: 4 bytes *)
+    `Server, "\000\003"; (* transmission flags: NBD_FLAG_READ_ONLY (bit 1) + NBD_FLAG_HAS_FLAGS (bit 0) *)
+    `Server, (String.make 124 '\000');
+    (* Now we've entered transmission mode *)
+
+    `Client, nbd_request_magic;
+    `Client, "\000\000"; (* command flags *)
+    `Client, "\000\000"; (* request type: NBD_CMD_READ *)
+    `Client, "\000\000\000\000\000\000\000\000"; (* handle: 4 bytes *)
+    `Client, "\000\000\000\000\000\000\000\001"; (* offset *)
+    `Client, "\000\000\000\002"; (* length *)
+
+    (* We're allowed to read from a read-only export *)
+    `Server, nbd_reply_magic;
+    `Server, "\000\000\000\000"; (* error: no error *)
+    `Server, "\000\000\000\000\000\000\000\000"; (* handle *)
+    `Server, "sd"; (* 2 bytes of data *)
+
+    `Client, nbd_request_magic;
+    `Client, "\000\000"; (* command flags *)
+    `Client, "\000\001"; (* request type: NBD_CMD_WRITE *)
+    `Client, "\000\000\000\000\000\000\000\001"; (* handle: 4 bytes *)
+    `Client, "\000\000\000\000\000\000\000\000"; (* offset *)
+    `Client, "\000\000\000\004"; (* length *)
+    (* The server should probably return the EPERM error immediately, and not
+       read any data associated with the write request, as the client should
+       recognize the error before transmitting the data, just like for EINVAL,
+       which is sent for unaligned requests. *)
+    (*`Client, "nope"; (* 4 bytes of data *)*)
+
+    (* However, we're not allowed to write to it *)
+    `Server, nbd_reply_magic;
+    `Server, "\000\000\000\001"; (* error: EPERM *)
+    `Server, "\000\000\000\000\000\000\000\001"; (* handle *)
+
+    (* TODO: currently the server disconnects in case of write errors, but
+       according to the NBD protocol it probably shouldn't, it should continue
+       to process the client's requests *)
+    (*
+    `Client, nbd_request_magic;
+    `Client, "\000\000"; (* command flags *)
+    `Client, "\000\002"; (* request type: NBD_CMD_DISC *)
+    `Client, "\000\000\000\000\000\000\000\002"; (* handle: 4 bytes *)
+    `Client, "\000\000\000\000\000\000\000\000"; (* offset *)
+    `Client, "\000\000\000\000"; (* length *)
+    *)
+  ]
+
```

Comment:
I take it this is to be uncommented once the `TODO` is implemented?

Reply:
Indeed, that's correct, it's commented because it's currently failing because of the TODO. I'll try to clarify this.